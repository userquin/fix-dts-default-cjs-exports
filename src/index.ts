import type { Options } from './utils'
import fs from 'node:fs/promises'
import { internalFixDefaultCJSExports } from './utils'

export type { Options }

export function transformDtsDefaultCJSExports(
  code: string,
  fileName: string,
  options: Options = {},
): string | undefined {
  return internalFixDefaultCJSExports(code, {
    fileName,
    // we don't need the imports (used only for exporting types optimization)
    imports: [],
  }, options)
}

export async function fixDtsFileDefaultCJSExports(
  dtsPath: string,
  options: Options = {},
): Promise<boolean> {
  const result = transformDtsDefaultCJSExports(
    await fs.readFile(dtsPath, 'utf-8'),
    dtsPath,
    options,
  )
  if (result) {
    await fs.writeFile(dtsPath, result, 'utf8')
  }
  return !!result
}

export interface TransformOptions {
  warn?: (message: string) => void
  transformLocalImports?: (
    code: string,
    dtsPath: string,
    dtsDestPath: string,
  ) => string
}

/**
 * Given an `ESM` dts file, transform it to a `d.ts` or `d.cts` file fixing CJS default exports.
 *
 * @param dtsPath The source `ESM` (`d.ts` or `.d.mts`) file path.
 * @param dtsDestPath The destination `.d.ts` or `.d.cts` file path.
 * @param options The options to use.
 * @see {@link defaultLocalImportsTransformer}
 */
export async function transformESMDtsToCJSDts(
  dtsPath: string,
  dtsDestPath: string,
  options: TransformOptions = {},
): Promise<void> {
  if (dtsPath === dtsDestPath) {
    throw new Error(`dtsPath and dtsDestPath should be different: ${dtsPath}`)
  }
  const code = await fs.readFile(dtsPath, 'utf-8')
  const result = transformDtsDefaultCJSExports(
    code,
    dtsPath,
    options,
  ) ?? code

  const { transformLocalImports = defaultLocalImportsTransformer } = options

  await fs.writeFile(
    dtsDestPath,
    transformLocalImports(dtsPath, dtsDestPath, result),
    'utf8',
  )
}

/**
 * Given an `ESM` dts file, transform it to a `d.ts` or `d.cts` file fixing CJS default exports.
 *
 * **NOTE**: local imports/exports will be replaced with the corresponding extension using source and destination files:
 * - when `dtsPath` is a `.d.ts` and `dtsDestPath` is a `d.cts`: `import { foo } from './foo.js'` -> `import { foo } from './foo.cjs'`
 * - when `dtsPath` is a `.d.mts` and `dtsDestPath` is a `d.ts`: `import { foo } from './foo.mjs'` -> `import { foo } from './foo.js'`
 * - when `dtsPath` is a `.d.mts` and `dtsDestPath` is a `d.cts`: `import { foo } from './foo.mjs'` -> `import { foo } from './foo.cjs'`
 *
 * @param code The code to transform.
 * @param dtsPath The source `ESM` (`d.ts` or `.d.mts`) file path.
 * @param dtsDestPath The destination `.d.ts` or `.d.cts` file path.
 */
export function defaultLocalImportsTransformer(
  code: string,
  dtsPath: string,
  dtsDestPath: string,
): string {
  // import { foo } from './foo.m?js'
  // import { foo } from '../foo.m?js'
  // export { foo } from './foo.m?js'
  // export { foo } from '../foo.m?js'
  const from = dtsPath.endsWith('.d.mts')
    ? /\s+(from\s+["'].\.?\/.*(\.mjs)["'];?)\s+/g
    : /\s+(from\s+["'].\.?\/.*(\.js)["'];?)\s+/g
  let matcher = from.exec(code)
  if (!matcher) {
    return code
  }
  const extension = dtsDestPath.endsWith('d.ts') ? '.js' : '.cjs'
  while (matcher) {
    code = code.replaceAll(matcher[1], matcher[1].replace(matcher[2], extension))
    matcher = from.exec(code)
  }
  return code
}
