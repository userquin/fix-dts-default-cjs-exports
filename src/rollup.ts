import type { Plugin, RenderedChunk } from 'rollup'
import { internalFixDefaultCJSExports } from './utils'

export interface PluginOptions {
  warn?: (message: string) => void
  matcher?: (info: RenderedChunk) => boolean
}

export function cjsExportsDtsMatcher(info: RenderedChunk): boolean {
  return info.type === 'chunk'
    && info.exports?.length > 0
    && info.exports.includes('default')
    && /\.d\.c?ts$/.test(info.fileName)
}

export function defaultCjsExportsDtsMatcher(info: RenderedChunk): boolean {
  return cjsExportsDtsMatcher(info) && info.isEntry
}

export function FixDtsDefaultCjsExportsPlugin(options: PluginOptions = {}): Plugin {
  const { matcher = defaultCjsExportsDtsMatcher } = options

  return {
    name: 'fix-dts-default-cjs-exports-plugin',
    renderChunk(code, info) {
      return matcher(info)
        ? internalFixDefaultCJSExports(code, info, options)
        : undefined
    },
  } satisfies Plugin
}
