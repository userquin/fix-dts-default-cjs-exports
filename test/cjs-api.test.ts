import type { ESMExport, StaticImport } from 'mlly'
import fs from 'node:fs/promises'
import path from 'node:path'
import { findExports, findStaticImports, findTypeExports, parseStaticImport } from 'mlly'
import { describe, expect, it } from 'vitest'
import { defaultLocalImportsTransformer, transformDtsDefaultCJSExports } from '../src'

type CodeInfo = [
  name: string,
  types: ESMExport[],
  exports: ESMExport[],
  content: string,
  imports: StaticImport[],
]

describe('api: node10 and Node16 Default Exports Types', () => {
  const root = path.resolve('./test/cjs-types-fixture')
  function resolveFile(name: string) {
    return path.resolve(root, name)
  }
  function extractInfo(content: string, name: string): CodeInfo {
    return [
      name,
      findTypeExports(content),
      findExports(content),
      content,
      findStaticImports(content),
    ]
  }

  it('api: mixed declarations', async () => {
    const mts = 'index.d.mts'
    const code = await fs.readFile(resolveFile(`mixed-declarations/dist/${mts}`), 'utf-8')
    let content = transformDtsDefaultCJSExports(code, mts)
    expect(content).toBeDefined()
    content = defaultLocalImportsTransformer(
      content!,
      mts,
      mts.replace(/\.d\.mts$/, '.d.cts'),
    )
    expect(content).toMatchSnapshot()
    const [name, types, exports] = extractInfo(content!, 'mixed-declarations')
    expect(exports).toHaveLength(0)
    expect(types).toHaveLength(1)
    expect(
      types.find(e => e.names.includes('default')),
      `${name} should not have a default export`,
    ).toBeUndefined()
  })
  it('api: re-Export Types', async () => {
    const files = await Promise.all([
      'all',
      'index',
      'types',
    ].map(async (name) => {
      name = resolveFile(`reexport-types/dist/${name}.d.mts`)
      const content = await fs.readFile(name, 'utf8')
      const transformed = transformDtsDefaultCJSExports(content, name)
      // types.d.mts should not be transformed
      return [
        name,
        transformed,
        defaultLocalImportsTransformer(
          transformed ?? content,
          path.basename(name),
          path.basename(name).replace(/\.d\.mts$/, '.d.cts'),
        ),
      ] as const
    }))
    for (const file of files) {
      const [name, transformed, useContent] = file
      expect(useContent, `${name} content should be defined`).toBeDefined()
      if (name.endsWith('types.d.mts')) {
        expect(transformed, `${name} transform should be undefined`).toBeUndefined()
      }
      else {
        expect(transformed, `${name} transform should be defined`).toBeDefined()
      }
      expect(useContent).toMatchSnapshot()
      const [_name, types, exports, _content, imports] = extractInfo(useContent, name)
      if (name.endsWith('types.d.mts')) {
        expect(exports).toHaveLength(0)
        expect(types).toHaveLength(1)
        expect(
          types.find(e => e.names.includes('default')),
          `${name} should not have a default export`,
        ).toBeUndefined()
      }
      else if (name.endsWith('index.d.mts')) {
        expect(exports).toHaveLength(2)
        expect(types).toHaveLength(0)
        expect(imports).toHaveLength(1)
        expect(
          exports.find(e => e.names.includes('default')),
          `${name} should not have a default export`,
        ).toBeUndefined()
        expect(useContent).toMatch('export = plugin')
      }
      else if (name.endsWith('all.d.mts')) {
        expect(exports).toHaveLength(2)
        expect(types).toHaveLength(0)
        expect(imports).toHaveLength(1)
        expect(
          exports.find(e => e.names.includes('default')),
          `${name} should not have a default export`,
        ).toBeUndefined()
        const defaultImport = parseStaticImport(imports[0])
        expect(defaultImport.defaultImport).toBe('_default')
        expect(useContent).toMatch(`export = ${defaultImport.defaultImport}`)
      }
    }
  })
  it('api: re-Export as default', async () => {
    const files = await Promise.all([
      'asdefault',
      'defaultclass',
      'index',
      'magicstringasdefault',
      'resolveasdefault',
    ].map(async (name) => {
      name = resolveFile(`reexport-default/dist/${name}.d.mts`)
      const content = await fs.readFile(name, 'utf8')
      const transformed = transformDtsDefaultCJSExports(content, name)
      return [
        name,
        transformed,
        defaultLocalImportsTransformer(
          transformed ?? content,
          path.basename(name),
          path.basename(name).replace(/\.d\.mts$/, '.d.cts'),
        ),
      ] as const
    }))
    for (const file of files) {
      const [name, content, useContent] = file
      expect(content, `${name} transform should be defined`).toBeDefined()
      expect(useContent).toMatchSnapshot()
      const [_name, types, exports, _content, imports] = extractInfo(useContent, name)
      if (name.match(/[\\/]asdefault.d.mts$/)) {
        expect(exports).toHaveLength(0)
        expect(types).toHaveLength(0)
        expect(imports).toHaveLength(1)
        expect(
          types.find(e => e.names.includes('default')),
          `${name} should not have a default export`,
        ).toBeUndefined()
        const defaultImport = parseStaticImport(imports[0])
        expect(defaultImport.namedImports?.resolve).toBeDefined()
        expect(content).toMatch(`export = resolve`)
      }
      else if (name.endsWith('index.d.mts')) {
        expect(exports).toHaveLength(1)
        expect(types).toHaveLength(0)
        expect(imports).toHaveLength(1)
        expect(
          exports.find(e => e.names.includes('default')),
          `${name} should not have a default export`,
        ).toBeUndefined()
        const defaultImport = parseStaticImport(imports[0])
        expect(defaultImport.defaultImport).toBe('MagicString')
        expect(content).toMatch(`export = ${defaultImport.defaultImport}`)
      }
      else if (name.endsWith('magicstringasdefault.d.mts')) {
        expect(exports).toHaveLength(0)
        expect(types).toHaveLength(0)
        expect(imports.filter(i => !!i.imports)).toHaveLength(1)
        expect(
          exports.find(e => e.names.includes('default')),
          `${name} should not have a default export`,
        ).toBeUndefined()
        const defaultImport = parseStaticImport(imports[0])
        expect(defaultImport.defaultImport).toBe('_default')
        expect(content).toMatch(`export = ${defaultImport.defaultImport}`)
      }
      else if (name.endsWith('resolvedasdefault.d.mts')) {
        expect(exports).toHaveLength(0)
        expect(types).toHaveLength(0)
        expect(imports.filter(i => !!i.imports)).toHaveLength(1)
        expect(
          exports.find(e => e.names.includes('default')),
          `${name} should not have a default export`,
        ).toBeUndefined()
        const defaultImport = parseStaticImport(imports[0])
        expect(defaultImport.defaultImport).toBe('resolve')
        expect(content).toMatch(`export = ${defaultImport.defaultImport}`)
      }
      else if (name.endsWith('defaultclass.d.mts')) {
        expect(exports).toHaveLength(0)
        expect(types).toHaveLength(0)
        expect(imports.filter(i => !!i.imports)).toHaveLength(0)
        expect(
          exports.find(e => e.names.includes('default')),
          `${name} should not have a default export`,
        ).toBeUndefined()
        expect(content).toMatch(`export = DefaultClass`)
      }
    }
  })
})
