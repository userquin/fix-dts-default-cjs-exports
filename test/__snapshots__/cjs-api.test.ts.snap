// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`api: node10 and Node16 Default Exports Types > api: mixed declarations 1`] = `
"interface A {
    name: string;
}
interface B {
    name: string;
}
interface C {
    name: string;
}
interface Options {
    a?: A;
    b?: B;
    c?: C;
}
declare function plugin(options?: Options): Options;

// @ts-ignore
export = plugin;
export type { A, B, C, Options };
"
`;

exports[`api: node10 and Node16 Default Exports Types > api: re-Export Types 1`] = `
"import _default from './index.cjs';
// @ts-ignore
export = _default;
export { ResolvedOptions } from './index.cjs';
export { A, B, CC as C, CC, Options } from './types.cjs';
"
`;

exports[`api: node10 and Node16 Default Exports Types > api: re-Export Types 2`] = `
"import { Options } from './types.cjs';
export { A, B, CC as C, CC } from './types.cjs';

interface ResolvedOptions extends Options {
    name: string;
}
declare function plugin(options?: Options): ResolvedOptions;

// @ts-ignore
export = plugin;
export { Options, type ResolvedOptions };
"
`;

exports[`api: node10 and Node16 Default Exports Types > api: re-Export Types 3`] = `
"interface A {
    name: string;
}
interface B {
    name: string;
}
interface C {
    name: string;
}

interface Options {
    a?: A;
    b?: B;
    c?: C;
}

export type { A, B, C, C as CC, Options };
"
`;

exports[`api: node10 and Node16 Default Exports Types > api: re-Export as default 1`] = `
"import { resolve } from 'pathe';
export = resolve;
"
`;

exports[`api: node10 and Node16 Default Exports Types > api: re-Export as default 2`] = `
"declare class DefaultClass {
    name: string;
    constructor(name: string);
}

export = DefaultClass;
"
`;

exports[`api: node10 and Node16 Default Exports Types > api: re-Export as default 3`] = `
"import MagicString from 'magic-string';
// @ts-ignore
export = MagicString;
export { MagicStringOptions } from 'magic-string';
"
`;

exports[`api: node10 and Node16 Default Exports Types > api: re-Export as default 4`] = `
"import _default from 'magic-string';
export = _default;
import 'pathe';
"
`;

exports[`api: node10 and Node16 Default Exports Types > api: re-Export as default 5`] = `
"import { resolve } from 'pathe';
export = resolve;
"
`;
